tools/cargos/src/main.rs

```diff
@@ -0,0 +1,254 @@
+    }
+
+    loop {
+        eprint!("[cargos] would you like to install toolchain '{name}' and component 'rust-src' via 'rustup' (y/n)? ");
+        let mut input = [0];
+        io::stdin().read_exact(&mut input).unwrap();
+        match input[0] as char {
+            'y' | 'Y' => break,
+            'n' | 'N' => process::exit(1),
+            _ => (),
+        }
+    }
+
+    let version = versions.get(name)?;
+    Command::new("rustup")
+        .args(["toolchain", "install", &version, "-c", "rust-src"])
+        .status()
+        .unwrap();
+
+    Ok(())
+}
+
+fn get_rustflags(name: &str) -> &'static str {
+    if name == "nightly" {
+        "--cfg __INTERNAL_USE_ONLY_NIGHTLY_FEATURES_IN_TESTS"
+    } else {
+        ""
+    }
+}
+
+fn rustup<'a>(args: impl IntoIterator<Item = &'a str>, env: Option<(&str, &str)>) -> Output {
+    let mut cmd = Command::new("rustup");
+    // It's important to set `RUSTUP_TOOLCHAIN` to override any value set while
+    // running this program. That variable overrides any `+<version>` CLI
+    // argument.
+    cmd.args(args).env("RUSTUP_TOOLCHAIN", "");
+    if let Some((name, val)) = env {
+        cmd.env(name, val);
+    }
+    cmd.output().unwrap()
+}
+
+fn delegate_cargo() -> Result<(), Error> {
+    let mut args = env::args();
+    let this = args.next().unwrap();
+    let versions = get_toolchain_versions();
+
+    match args.next().as_deref() {
+        None => Err(Error::NoArguments),
+        Some("--version") => {
+            let name = args.next().ok_or(Error::MissingToolchainVersion)?;
+            println!("{}", versions.get(&name)?);
+            Ok(())
+        }
+        Some("+all") => {
+            eprintln!("[cargos] warning: running the same command for each toolchain (msrv, stable, nightly)");
+            let args = args.collect::<Vec<_>>();
+
+            for toolchain in ["msrv", "stable", "nightly"] {
+                eprintln!("[cargos] running with toolchain: {toolchain}");
+                Command::new(this.clone())
+                    .arg(format!("+{toolchain}"))
+                    .args(args.clone())
+                    .status()
+                    .unwrap();
+            }
+            Ok(())
+        }
+        Some(arg) => {
+            if let Some(name) = arg.strip_prefix('+') {
+                let version = versions.get(name)?;
+
+                if !is_toolchain_installed(&versions, name)? {
+                    install_toolchain_or_exit(&versions, name)?;
+                }
+
+                let rustflags = env::vars()
+                    .filter_map(|(k, v)| if k == "RUSTFLAGS" { Some(v) } else { None })
+                    .next()
+                    .unwrap_or_default();
+
+                Command::new("rustup")
+                    .args(["run", &version, "cargo"])
+                    .args(args)
+                    .env("RUSTFLAGS", format!("{}{}", get_rustflags(name), rustflags))
+                    .status()
+                    .unwrap();
```

Comment:
Could be worth factoring this out, but it can't be `fn rustup` (at least as it's written now) because that doesn't wire up stdin/stdout/stderr to the parent process's stdin/stdout/stderr.

Reply:
Done